text/0000-rustdoc-json.md

```diff
@@ -0,0 +1,1124 @@
+
+- [ClangDoc] has the ability to output either rendered HTML, or tool consumable YAML.
+- [PureScript] uses an intermediate JSON representation when publishing package information to their
+  [Pursuit] directory. It's primarily used to generate documentation, but can also be used to
+  generate `etags` files.
+- [DartDoc] is in the process of implementing a JSON ouput.
+- [Doxygen] has an option to generate an XML file with the code's information.
+- [Haskell]'s documentation tool, [Haddock], can generate an intermediate representation used by the
+  type search engine [Hoogle] to integrate documentation of several packages.
+- [Kythe] is a "(mostly) language-agnostic" system for integrating documentation across several
+  langauges. It features its own schema that code information can be translated into, that services
+  can use to aggregate information about projects that span multiple languages.
+- [GObject Introspection] has an intermediate XML representation called GIR that's used to create
+  langauge bindings for GObject-based C libraries. While (at the time of this writing) it's not
+  currently used to create documentation, it is a stated goal to use this information to document
+  these libraries.
+
+[ClangDoc]: https://clang.llvm.org/extra/clang-doc.html/
+[PureScript]: http://www.purescript.org/
+[Pursuit]: https://pursuit.purescript.org/
+[DartDoc]: https://dart.dev/tools/dartdoc/
+[Doxygen]: https://www.doxygen.nl/
+[Haskell]: https://www.haskell.org/
+[Haddock]: https://www.haskell.org/haddock/
+[Hoogle]: https://www.haskell.org/hoogle/
+[Kythe]: http://kythe.io/
+[GObject Introspection]: https://gi.readthedocs.io/en/latest/
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What is the stabilization story? As language features are added, this representation will need to
+  be extended to accommodate it. As this will change the structure of the data, what does that mean
+  for its consumers?
+- How will users be able to manipulate the data? Is it a good idea to host a crate outside the
+  compiler that contains the struct definitions for all the types that get serialized so that
+  people could easily hack on the data without the compiler? Should that crate be the source of
+  truth for those types and be depended on by librustdoc, or should it be a mirror that gets
+  updated externally to reflect the changes to the copy in the compiler?
+- How will intra-doc links be handled?
+  - Supporting `struct.SomeStruct.html` style links seems infeasible since it would tie alternative
+    front-ends to `rustdoc`'s file/folder format.
+  - With the nightly [intra-rustdoc link syntax](https://github.com/rust-lang/rust/pull/47046) it's
+    debatable whether we should resolve those to HTML links or leave that up to whatever consumes
+    the JSON. Leaving them unresolved seems preferable but it would mean that consumers have to do
+    markdown parsing to replace them with actual links.
```

Comment:
It is not possible in general to resolve intra-doc links with the format specified in this RFC. That requires knowing which paths are in scope where the item is documented, not all of which are documented or public.

Reply:
I didn't think about the fact that being able to write unqualified paths means that scoping info is required. Maybe it would be possible to resolve them to their fully qualified module paths and modify the markdown to output those (`"docstring referencing [Iterator]"` -> `"docstring referencing [Iterator](trait@foo::Iterator)"` or something)? Alternatively we could have another field with the list of IDs or fully qualified paths corresponding to each link in the markdown block in order.

Reply:
> `trait@foo::Iterator`

This is still ambiguous, because you could have multiple crates with the same name, e.g. `tokio 0.1` and `tokio 0.2`.

Reply:
Maybe an alternative is to have intra-doc links as a separate output field? Something like
```json
{
    "intra-doc-links": {
        "struct@foo::Iterator": 0,
   }
}
```
where the value is the ID. That assumes the linked item is part of the output though, which isn't always true for cross-crate re-exports.

Reply:
In that case how about we include a list of IDs corresponding to the resolved items from the intradoc links in the docstring. That way the consumer can use those to look up the items in `path` or `index` and generate links to them using their `crate_num` which solves the versioning issue.